lib/spack/spack/directives.py

```diff
@@ -248,8 +249,14 @@ def _execute_version(pkg: Type[spack.package_base.PackageBase], ver: Union[str,
             f"    version('{ver}', {args})"
         )
 
-    # Store kwargs for the package to later with a fetch_strategy.
-    pkg.versions[version] = kwargs
+    if spack.mirrors.utils.get_mirror_all_mode() and kwargs.get("url"):
+        hash_val = next((kwargs[s] for s in spack.util.crypto.hashes if s in kwargs), None)
+        if hash_val:
+            version_obj = Version(f"{version}.{hash_val[:16]}")
+            pkg.versions[version_obj] = kwargs
+    else:
+        # Store kwargs for the package to later with a fetch_strategy.
```

Comment:
I think this comment can be removed (I see it predates this PR, but I'm not sure what it adds, or what it means given that it is incomplete).